### ISSUE TITLE
Provide & install a pkg-config file                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

### ISSUE THREAD

**Issue by ingmarv** [weidai11/cryptopp#133]
Any plans for installing a .pc along with cryptopp?

Right now I see the following:
- distributions like Debian install their own, including maintaining an autoconf buildsystem to do that.
- master contains a cmake buildsystem, that installs a .cmake, which helps packages using cmake find cryptopp, but doesn't help any others

**Comment by noloader** [comment]
> Any plans for installing a .pc along with cryptopp?

No.

-----

> distributions like Debian install their own, including maintaining an autoconf buildsystem to do that.

We offered to provide one for distos, but there was no interest from them. The distros we reached out to included Debian, Fedora, Ubuntu, Macports, Gentoo and Arch Linux.

We emailed the distros asked them for feedback on the mailing list at [Distro maintainers: pkgconfig nad DataDir patch?](http://groups.google.com/d/msg/cryptopp-users/VqPmjsxJBVE/U7u1cqjf9PoJ). There were no replies.

We then opened a bug report for distros to comment at [Crypto++ should provide a pkg-config for users and distros](http://github.com/weidai11/cryptopp/issues/36). Again, there were no replies.

------

> master contains a cmake buildsystem, that installs a .cmake, which helps packages using cmake find cryptopp, but doesn't help any others

Yeah, that's just a makefile generator that was intended to help build the library. It was also intended to help port to new platforms like Windows, Windows Phone, ARMHF, ARM Neon, etc. Based on testing and bug reports, it appears to be mostly non-working.

I think we were a bit premature in cutting Cmake in. We should have placed it on the [Patch Page](https://www.cryptopp.com/wiki/Category:Patch) as an optional content until it matured.

-----

There's a not-so-readily-apparent issue with a pkg-config. The problem is the Crypto++ library goes by different names on different platforms. On Debian, Ubuntu and derivatives, the library is called ***`libcrypto++.{a|so}`***; while on Fedora and friends its called ***`libcryptopp.{a|so}`***. As far as we know, there is no way to abstract that in a pkg-config file. Also see [Abstracting include paths and library names with pkg-config?](http://stackoverflow.com/q/32920396) on Stack Overflow.

-----

Feel free to add a pkg-config and place it at the [Patch Page](https://www.cryptopp.com/wiki/Category:Patch) as an optional component. The patch page is used as an incubator or an area where stuff that does not quite fit in the library can reside. This seems like a good candidate for the page.

Also be aware that we provide an implicit blessing of stuff that shows up at [Patch Page](https://www.cryptopp.com/wiki/Category:Patch). We even incorporate some of them in our testing phase. The patch page's goals and objectives are:

* Provide things that the library lacks but may help library users
* Provide them under the same terms of the library so its easy to uptake

Please be sure to respect licensing requirements, and don't use, say, one of GNU's restrictive licenses that forces users to do things. If you feel inclined to provide a restrictive GNU license, then please avoid the patch pages. Instead, place it in your personal corner of the web and let search engines lead folks to it.

**Comment by noloader** [comment]
Duplicate of [Issue 36, Crypto++ should provide a pkg-config for users and distros](https://github.com/weidai11/cryptopp/issues/36)

**Comment by ingmarv** [comment]
> The problem is the Crypto++ library goes by different names on different platforms.

I fail to see how that's your problem. They break it they get to keep both pieces.
In the mean time, it'd be handy for distros who try to follow what upstream does to have a .pc file installed. I'll see if if I find the time for a patch.

Thanks for responding.

**Comment by Chemrat** [comment]
> There's a not-so-readily-apparent issue with a pkg-config. The problem is the Crypto++ library goes by different names on different platforms. On Debian, Ubuntu and derivatives, the library is called libcrypto++.{a|so}; while on Fedora and friends its called libcryptopp.{a|so}.

It seems like a problem with Debian. GNUmakefile and CMakeLists.txt provided in master tree build libcryptopp.so/libcryptopp.a and that's probably the only pkg-config should be provided. If downstream decides to rename the package, then downstream should fix the .pc file as well?

Also while Debian bundles the library as crypto++, it also provides a cryptopp symlink for headers and libraries (see Debian bug [Please support the upstream names for include files and libraries](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=448766) and Debian's [libcrypto++-dev.links.in](https://sources.debian.net/src/libcrypto%2B%2B/5.6.3-6/debian/libcrypto%2B%2B-dev.links.in/))

Despite apparent lack of interest from distros, it's still a problem because different distros bundling different .pc files make it troublesome to build software using crypto++ across different distros (e.g. clementine-player/Clementine#5279). Standardized pkg-config file from the upstream would definitely help.

**Comment by Polynomial-C** [comment]
Hi,

this just bit me again in [Gentoo bug #654572](https://bugs.gentoo.org/654572) and I wonder how often this might happen to others as well (not only with clementine of course).
Yes I know you gave distributions the opportunity to comment and help establish a pkg-config file already but honestly, being a Gentoo developer, if I had seen your inquiry back then I most certainlay would have answered and tried to help.
I still think having an official pkg-config file from upstream would ease things very much. And with the different library names... well I suppose that can be managed to still work with pkg-config file.
I sincerely hope you are going to reconsider your decision at some point. At lest you'd make one package maintainer a bit more happy :-)

**Comment by Polynomial-C** [comment]
Oh, and we already had such a request at Gentoo twice:

[Gentoo bug #505550](https://bugs.gentoo.org/505550)
[Gentoo bug #548542](https://bugs.gentoo.org/548542)

So it seems this was more like a communication issue. Do you know your inquiry really arrived at the parties you wanted to involve?

**Comment by noloader** [comment]
@Polynomial-C,

The makefile has a recipe to make a PC file:

```
$ make libcryptopp.pc
$ cat libcryptopp.pc
# Crypto++ package configuration file

prefix=/usr/local
libdir=/usr/local/lib
includedir=${prefix}/include

Name: Crypto++
Description: Crypto++ cryptographic library
Version: 7.1
URL: https://cryptopp.com/

Cflags: -I${includedir}
Libs: -L${libdir} -lcryptopp
```

There's also an Autotools project at [Noloader | Autotools](https://github.com/noloader/cryptopp-autotools) that may be helpful for you.

**Comment by Polynomial-C** [comment]
Even better :)
I've filed [Gentoo bug #654718](https://bugs.gentoo.org/654718).
Thanks for your reply.

**Comment by noloader** [comment]
No problems @Polynomial-C.

For completeness, the PC makefile recipe was added recently. I think it was first available with Crypto++ 6.0 which was released in February 2018.

We don't really advertise it because it is kind of fragile. The name may be wrong, it does not set RPATH's, it uses `CFLAGS` instead of `CXXFLAGS`, etc. There's a lot of problems that the pkg-config folks need to work out.
